### PR TITLE
[miio] Allow for custom refresh commands

### PR DESCRIPTION
**Pull request by marcelrv**
* Allow for custom commands during miio:basic refreshes, needed to
support refresh for e.g. scicoffeemaker, viomi vacuums etc

* Improve the unsupported handler to deal with the custom commands as
well

* Do not add non-existing channeltypes in basic handler

* Add the right channeltypes to the jsonDB instead of just defaulting
them to the channel names.

Signed-off-by: Marcel Verpaalen marcel@verpaalen.com

### CODE REVIEW

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoBasicHandler.java [hunk 1]
```diff
@@ -83,12 +83,16 @@
     });
 
     List<MiIoBasicChannel> refreshList = new ArrayList<>();
+    Map<String, MiIoBasicChannel> refreshListCustomCommands = new HashMap<>();
```
Comment (cpmeister) [inline bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoBasicHandler.java:86]:
```suggestion
    private Map<String, MiIoBasicChannel> refreshListCustomCommands = new HashMap<>();
```

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoBasicHandler.java [hunk 2]
```diff
@@ -398,17 +415,24 @@ private boolean buildChannelStructure(String deviceName) {
...
             logger.info("Channel '{}' for thing {} already exist... removing", channel, getThing().getUID());
             thingBuilder.withoutChannel(new ChannelUID(getThing().getUID(), channel));
         }
-        Channel newChannel = ChannelBuilder.create(channelUID, datatype).withType(channelTypeUID)
-                .withLabel(friendlyName).build();
-        thingBuilder.withChannel(newChannel);
+        ChannelBuilder newChannel = ChannelBuilder.create(channelUID, datatype).withLabel(friendlyName);
+        if (!channelType.isBlank()) {
+            ChannelTypeUID channelTypeUID = new ChannelTypeUID(channelType);
+            if (channelTypeRegistry.getChannelType(channelTypeUID) != null) {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoBasicHandler.java:428]: I'd imagine that channel types are contextual to the bindings that they are defined in, without the binding prefix I don't think that the registry would ever find the channel type you are looking for other than the system ones.

**Review by cpmeister — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoBasicHandler.java (hunk 2 above)
Comment (marcelrv) [reply bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoBasicHandler.java:428]: True, but the binding prefix is added in the the getChannelType

https://github.com/openhab/openhab-addons/pull/8796/files#diff-1fbdf89cc35d86f5da2024be7a5588a90fca7e7314551a250755469e435ac900L138

Comment (cpmeister) [reply bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoBasicHandler.java:428]: Ah, I saw the part you removed, but didn't notice the part you left in. You are correct.

File: bundles/org.openhab.binding.miio/src/main/resources/database/chuangmi.plug.v3fw.json [hunk 3]
```diff
@@ -1,64 +1,68 @@
...
+        "actions": [
+          {
+            "command": "set_power",
+            "parameterType": "ONOFF"
+          }
+        ],
+        "readmeComment": "If this channel does not respond to on/off replace the model with chuangmi.plug.v3old in the config or upgrade firmware"
+      },
+      {
+        "property": "usb_on",
```
Comment (cpmeister) [inline bundles/org.openhab.binding.miio/src/main/resources/database/chuangmi.plug.v3fw.json:27]: Also, why did the formatting change for this file? It is not inconsistent with the other files.

Comment (marcelrv) [reply bundles/org.openhab.binding.miio/src/main/resources/database/chuangmi.plug.v3fw.json]: I used that file for my testing, and did some intrim reset to undo my changes for the testing.
seems that that also undone the formatting changes...
Applied the formatter once more.

**Review by cpmeister — APPROVED** [review]
LGTM

**Pull request merged**